webapi/payfee.go

```diff
@@ -128,42 +124,47 @@ func payFee(c *gin.Context) {
 		return
 	}
 
-	// Loop through transaction outputs until we find one which pays to the
-	// expected fee address. Record how much is being paid to the fee address.
-	var feePaid dcrutil.Amount
-	const scriptVersion = 0
+	// Decode fee address to get its payment script details.
+	feeAddr, err := stdaddr.DecodeAddress(ticket.FeeAddress, cfg.NetParams)
+	if err != nil {
+		log.Errorf("%s: Failed to decode fee address (ticketHash=%s): %v",
+			funcName, ticket.Hash, err)
+		sendError(errInternalError, c)
+		return
+	}
+
+	wantScriptVer, wantScript := feeAddr.PaymentScript()
 
-findAddress:
+	// Confirm the provided fee transaction contains an output which pays to the
+	// expected payment script. Both script and script version should match.
+	var feePaid dcrutil.Amount
 	for _, txOut := range feeTx.TxOut {
-		if txOut.Version != scriptVersion {
-			log.Errorf("%s: Fee tx with invalid script version (clientIP=%s, ticketHash=%s): was %d, expected %d",
-				funcName, c.ClientIP(), ticket.Hash, txOut.Version, scriptVersion)
-			sendErrorWithMsg("invalid script version", errInvalidFeeTx, c)
-			return
-		}
-		_, addresses, _, err := txscript.ExtractPkScriptAddrs(scriptVersion,
-			txOut.PkScript, cfg.NetParams, isTreasuryEnabled)
-		if err != nil {
-			log.Errorf("%s: Extract PK error (clientIP=%s, ticketHash=%s): %v",
-				funcName, c.ClientIP(), ticket.Hash, err)
-			sendError(errInternalError, c)
-			return
-		}
-		for _, addr := range addresses {
-			if addr.String() == ticket.FeeAddress {
-				feePaid = dcrutil.Amount(txOut.Value)
-				break findAddress
-			}
+		if txOut.Version == wantScriptVer && bytes.Equal(txOut.PkScript, wantScript) {
+			feePaid = dcrutil.Amount(txOut.Value)
+			break
 		}
 	}
 
+	// Confirm a fee payment was found.
```

Comment:
Unless you want to specifically distinguish the errors in the case of no payment vs paying too little, you could remove this check altogether since the next minimum fee check would catch it too, assuming `ticket.FeeAmount` is already prevented from being zero previously, of course.

Reply:
There are different error codes for incorrect fee amount vs no fee, hence the seperate checks